File: java/L100/LeetCode0105.java
package L100;

import util.TreeNode;

public class LeetCode0105 {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        if (preorder.length == 0) {
            return null;
        }
        return help(preorder, 0, preorder.length - 1, inorder, 0, inorder.length - 1);
    }
    public TreeNode help(int[] preorder,int pi, int pj, int[] inorder, int ii, int ij) {
        if (pi > pj) {
            return null;
        }
        TreeNode cur = new TreeNode(preorder[pi]);
        if (pi == pj) {
            return cur;
        }
        int iroot = ii;
        for (;iroot <= ij;++iroot) {
            if (inorder[iroot] == preorder[pi]) {
                break;
            }
        }
        TreeNode left = help(preorder, pi + 1, pi + iroot - ii, inorder, ii, iroot - 1);
        TreeNode right = help(preorder, pi + iroot - ii + 1, pj, inorder, iroot + 1, ij);
        cur.left = left;
        cur.right = right;
        return cur;
    }
}

File: cpp/L100/LeetCode0105.cc
#include "../heads/tree_node.h"
#include <utility>
#include <vector>
using std::vector;

class Solution {
    int findIndex(vector<int> &inorder, int ilow, int iHigh, int value) {
        for (int i = ilow;i <= iHigh;++i) {
            if (inorder[i] == value) {
                return i;
            }
        }
        return -1;
    }
    std::pair<TreeNode*, int> buildChildTree(vector<int>& preorder, int plow, int phigh, vector<int>& inorder, int ilow, int iHigh) {
        if (plow > phigh || ilow > iHigh) {
            return {nullptr, plow};
        }
        TreeNode *node = new TreeNode(preorder[plow]);
        int id = findIndex(inorder, ilow, iHigh, preorder[plow]);
        auto [ln, p1] = buildChildTree(preorder, plow + 1, phigh, inorder, ilow, id - 1);
        auto [rn, p2] = buildChildTree(preorder, p1, phigh, inorder, id + 1, iHigh);
        node->left = ln;
        node->right = rn;
        return {node, p2};
    }
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        int pN = preorder.size(), iN = inorder.size();
        auto [node, id] = buildChildTree(preorder, 0, pN - 1, inorder, 0, iN - 1);
        return node;
    }
};